### DISCUSSION TITLE
How to put different OCV curve during charge and discharge

### DISCUSSION THREAD

**Discussion by ashish17khandelwal** [pybamm-team/PyBaMM#1993] [General]
I want to put different open cell voltage table for charge and discharge (for each electrode).
I understand I can make a py function, but I am stuck on how to get a flag variable of charge/ discharge

**Comment by brosaplanella** [comment]
I am not sure how you could do this as a function because the current is not an input of the OCV so inside the function you can't know if you are charging or discharging. If you have clearly defined charge-discharge cycles, you could run half a cycle, swap the parameters, and run the next half initialising the model with the previous solution.

**Comment by TomTranter** [comment]
Could you work around it by including another overpotential that depends on current sign then use the average OCV and add or subtract the difference? @rtimms has done some work on hysteresis and may have some thoughts

**Reply by rtimms** [reply]
If you are directly prescribing the current as a parameter you can just use the “Current function [A]” parameter directly in the definition of the OCP like in [this example](https://github.com/weilongai/PyBaMM/blob/53e5ab00155b151d16797878e84ae4efe42e6b45/pybamm/input/parameters/lithium_ion/negative_electrodes/graphite_Chen2020_composite/silicon_ocp_Mark2016.py). This won’t work when you aren’t prescribing the current directly (e.g. it won’t work for experiments where you give the voltage or power). 

I think the best solution would be to make an OCP submodel. The default one would just set the OCV using the parameter like we currently do, but it would be relatively straightforward to add other submodels to account for hysteresis. These could be as straightforward as providing an OCP for charge and discharge and switching between the two, or you could add extra equations for new variables that control the hysteresis (e.g. like the models in Plett’s book).

**Reply by valentinsulzer** [reply]
OCP submodel is a good idea
